src/network/netdev/macsec.c

```diff
@@ -0,0 +1,1216 @@
+        if (r < 0)
+                return r;
+
+        r = netlink_call_async(netdev->manager->genl, NULL, m, macsec_receive_association_handler,
+                               netdev_destroy_callback, netdev);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Failed to configure receive secure association: %m");
+
+        netdev_ref(netdev);
+
+        return 0;
+}
+
+static int netdev_macsec_configure_receive_channel(NetDev *netdev, MACsecSCI *sci) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        int r;
+
+        assert(netdev);
+        assert(sci);
+
+        r = netdev_macsec_fill_message(netdev, MACSEC_CMD_ADD_RXSC, &m);
+        if (r < 0)
+                return r;
+
+        r = netdev_macsec_fill_message_sci(netdev, sci, m);
+        if (r < 0)
+                return r;
+
+        r = netlink_call_async(netdev->manager->genl, NULL, m, macsec_receive_channel_handler,
+                               netdev_destroy_callback, netdev);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Failed to configure receive channel: %m");
+
+        netdev_ref(netdev);
+
+        return 0;
+}
+
+static int netdev_macsec_configure_transmit_association(NetDev *netdev, TransmitAssociation *a) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        int r;
+
+        assert(netdev);
+        assert(a);
+
+        r = netdev_macsec_fill_message(netdev, MACSEC_CMD_ADD_TXSA, &m);
+        if (r < 0)
+                return r;
+
+        r = netdev_macsec_fill_message_sa(netdev, &a->sa, m);
+        if (r < 0)
+                return r;
+
+        r = netlink_call_async(netdev->manager->genl, NULL, m, macsec_transmit_association_handler,
+                               netdev_destroy_callback, netdev);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Failed to configure transmit secure association: %m");
+
+        netdev_ref(netdev);
+
+        return 0;
+}
+
+static int netdev_macsec_configure(NetDev *netdev, Link *link, sd_netlink_message *m) {
+        ReceiveAssociation *n;
+        TransmitAssociation *a;
+        ReceiveChannel *c;
+        Iterator i;
+        MACsec *s;
+        int r;
+
+        assert(netdev);
+
+        s = MACSEC(netdev);
+
+        assert(s);
+
+        ORDERED_HASHMAP_FOREACH(c, s->receive_channel_by_section, i) {
+                r = netdev_macsec_configure_receive_channel(netdev, &c->sci);
+                if (r < 0)
+                        return r;
+        }
+
+        ORDERED_HASHMAP_FOREACH(a, s->transmit_association_by_section, i) {
+                r = netdev_macsec_configure_transmit_association(netdev, a);
+                if (r < 0)
+                        return r;
+        }
+
+        ORDERED_HASHMAP_FOREACH(n, s->receive_association_by_section, i) {
+                r = netdev_macsec_configure_receive_channel(netdev, &n->sci);
```

Comment:
We can move to the call back of ```netdev_macsec_configure_transmit_association``` .

Reply:
??? Sorry, I cannot catch what you want to say. Could you elaborate more?

Reply:
sorry ```netdev_macsec_configure_receive_channel```'s callback that is ```macsec_receive_channel_handler```. That way it also makes sure that receive channel is always created first.

Reply:
Ah, I see. Thanks.

Reply:
In the revised version, channels are created at first, and relevant SAs are sent in the callback for each channel.

Reply:
yes that will work.